 MACHINE LEARNING PROJECT 

 TOPIC :**Multiple Linear Regression**
***

Exploratory Data Analysis 

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import datasets
import statsmodels.api as sm
import statsmodels.formula.api as smf

In [ ]:
df=pd.read_csv("/content/Lab 2 dataset ML.csv")
df.head()

S_no  Sales  CompPrice  Income  Advertising  Population  Price ShelveLoc  \
0     1   9.50        138      73           11         276    120       Bad   
1     2  11.22        111      48           16         260     83      Good   
2     3  10.06        113      35           10         269     80    Medium   
3     4   7.40        117     100            4         466     97    Medium   
4     5   4.15        141      64            3         340    128       Bad   

   Age  Education Urban   US  
0   42         17   Yes  Yes  
1   65         10   Yes  Yes  
2   59         12   Yes  Yes  
3   55         14   Yes  Yes  
4   38         13   Yes   No

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   S_no         400 non-null    int64  
 1   Sales        400 non-null    float64
 2   CompPrice    400 non-null    int64  
 3   Income       400 non-null    int64  
 4   Advertising  400 non-null    int64  
 5   Population   400 non-null    int64  
 6   Price        400 non-null    int64  
 7   ShelveLoc    400 non-null    object 
 8   Age          400 non-null    int64  
 9   Education    400 non-null    int64  
 10  Urban        400 non-null    object 
 11  US           400 non-null    object 
dtypes: float64(1), int64(8), object(3)
memory usage: 37.6+ KB


In [ ]:
df=df.drop(["S_no"],axis=1)

In [ ]:
df.isnull().sum()

Sales          0
CompPrice      0
Income         0
Advertising    0
Population     0
Price          0
ShelveLoc      0
Age            0
Education      0
Urban          0
US             0
dtype: int64

In [ ]:
df.describe()

Sales   CompPrice      Income  Advertising  Population  \
count  400.000000  400.000000  400.000000   400.000000  400.000000   
mean     7.496325  124.975000   68.657500     6.635000  264.840000   
std      2.824115   15.334512   27.986037     6.650364  147.376436   
min      0.000000   77.000000   21.000000     0.000000   10.000000   
25%      5.390000  115.000000   42.750000     0.000000  139.000000   
50%      7.490000  125.000000   69.000000     5.000000  272.000000   
75%      9.320000  135.000000   91.000000    12.000000  398.500000   
max     16.270000  175.000000  120.000000    29.000000  509.000000   

            Price         Age   Education  
count  400.000000  400.000000  400.000000  
mean   115.795000   53.322500   13.900000  
std     23.676664   16.200297    2.620528  
min     24.000000   25.000000   10.000000  
25%    100.000000   39.750000   12.000000  
50%    117.000000   54.500000   14.000000  
75%    131.000000   66.000000   16.000000  
max    191.000000   80.000000   18.000000

In [ ]:
df.describe(exclude='number') # exclude numerical columns


ShelveLoc Urban   US
count        400   400  400
unique         3     2    2
top       Medium   Yes  Yes
freq         219   282  258

In [ ]:
df_urban=pd.get_dummies(df["Urban"])
df_us=pd.get_dummies(df["US"])
df["Urban"]=df_urban["Yes"]
df["US"]=df_us["Yes"]

In [ ]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.3)


Model Fitting

In [ ]:
from sklearn.linear_model import LinearRegression

model = LinearRegression()


In [ ]:
model.fit(x_train,y_train)

LinearRegression()

In [ ]:
print(model.coef_)
print(model.intercept_)


[-0.05765695  1.2669841   0.1936657 ]
13.243728789316737


In [ ]:
a=[x_train,y_train]
df1=pd.concat(a , axis=1)

In [ ]:
mod_1 = smf.ols(formula = "Sales ~ Price + Urban + US", data = df1)
res_1 = mod_1.fit()
print(res_1.summary())

                            OLS Regression Results                            
Dep. Variable:                  Sales   R-squared:                       0.274
Model:                            OLS   Adj. R-squared:                  0.267
Method:                 Least Squares   F-statistic:                     34.80
Date:                Thu, 16 Mar 2023   Prob (F-statistic):           4.15e-19
Time:                        07:59:59   Log-Likelihood:                -647.21
No. Observations:                 280   AIC:                             1302.
Df Residuals:                     276   BIC:                             1317.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     13.2437      0.749     17.674      0.0

### Results


The coefficient of -0.054459 for Price means that, for a given location (i.e. fixed values of Urban and US), increasing the price of a car seat by $1 results in a decrease of sales by approximately 54.46 units, on average, in the model. The coefficient of -0.021916 for UrbanYes means that, for a given carseat price point and value of US, the model predicts urban areas to have approximately 22 fewer carseat sales on average compared to non-urban areas. The coefficient of 1.200573 for USYes means that, for a given carseat price point and value of Urban, the model predicts that stores in the United States have 1201 more carseat sales on average than stores outside the United States.

In [ ]:
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error


In [ ]:
predicted_train=model.predict(x_train)

In [ ]:
score=r2_score(y_train,predicted_train)
score

0.27446290214088587

In [ ]:
mean_sq=mean_squared_error(y_train,predicted_train)
mean_sq

5.959984591487779

In [ ]:
predicted_test=model.predict(x_test)

In [ ]:
error=predicted-y_test

In [ ]:
score=r2_score(y_test,predicted_test)
score

0.1382464825796117

In [ ]:
mean_sq=mean_squared_error(y_test,predicted_test)
mean_sq

6.321407283611482

p value for urban is 0.77 which is greater than level of significance 0.05. Hence the variable "Urban" is not significant. 

We will fit new model removing insignificant variable i.e "Urban" from the model.

In [ ]:
model1=LinearRegression()
model1.fit(x_train[["Price","US"]],y_train)

LinearRegression()

In [ ]:
print(model1.coef_)
print(model1.intercept_)


[-0.05747904  1.26927868]
13.361239329107391


In [ ]:
predicted_train1=model1.predict(x_train[["Price","US"]])

In [ ]:
score1_train=r2_score(y_train,predicted_train1)
score1_train

0.2735479075783982

In [ ]:
mean_sq1_train=mean_squared_error(y_train,predicted_train1)
mean_sq1_train

5.967500890116496

In [ ]:
predicted_test1=model1.predict

In [ ]:
score1_test=r2_score(y_test,predicted_test1)
score1_test

0.14498949082208257

In [ ]:
mean_sq1_test=mean_squared_error(y_test,predicted_test1)
mean_sq1_test

6.271943834312193

##### Conclusion:
 R^2 for the second model is more than that of 1st model. Also mean square error for second model is less than that of first model. Hence 2nd model is better fit for the data.

In [ ]:
mod_2 = smf.ols(formula = "Sales ~ Price + US", data = df1)
res_2 = mod_2.fit()
print(res_1.summary())

                            OLS Regression Results                            
Dep. Variable:                  Sales   R-squared:                       0.274
Model:                            OLS   Adj. R-squared:                  0.268
Method:                 Least Squares   F-statistic:                     52.15
Date:                Thu, 16 Mar 2023   Prob (F-statistic):           5.99e-20
Time:                        08:14:50   Log-Likelihood:                -647.39
No. Observations:                 280   AIC:                             1301.
Df Residuals:                     277   BIC:                             1312.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     13.3612      0.722     18.518      0.0

#####The confidence interval for intercept is (11.941,14.782)
##### The confidence interval for Price is (-0.069,0.046)
##### The confidence interval for US is (0.069,1.869)

The linear model has the form Y=β0+β1X1+β2X2+ϵ, where the regression coefficients are  β0=13.3612
 ,  β1=-0.0575, β2= 1.2693